src/data_chain/vote_accumulator.rs

```diff
@@ -0,0 +1,366 @@
+    fn create_full_infos() -> Vec<FullInfo> {
+        let mut rng = SeededRng::thread_rng();
+        unwrap!(rust_sodium::init_with_rng(&mut rng));
+
+        let mut nodes: Vec<_> = (0..GROUP_SIZE).map(|_| FullInfo::node_new(0)).collect();
+        nodes.sort_by(|lhs, rhs| lhs.public_info().cmp(rhs.public_info()));
+        for (index, node) in nodes.iter_mut().enumerate() {
+            node.set_age(index as u8 + 5);
+        }
+        nodes
+    }
+
+    // Create a block.  If `fill` is true, add votes for every member of `nodes`, otherwise only add
+    // votes until the block is valid.
+    fn create_block<T: Serialize + Clone>(payload: T, nodes: &[FullInfo], fill: bool) -> Block<T> {
+        let valid_voters = nodes
+            .iter()
+            .map(|full_info| *full_info.public_info())
+            .collect();
+        let mut vote = unwrap!(Vote::new(nodes[0].secret_sign_key(), payload.clone()));
+        let mut block = unwrap!(Block::new(&vote, nodes[0].public_info()));
+        for node in &nodes[1..] {
+            vote = unwrap!(Vote::new(node.secret_sign_key(), payload.clone()));
+            if let BlockState::Valid = unwrap!(
+                block.add_vote(&vote, node.public_info(), &valid_voters)
+            )
+            {
+                if !fill {
+                    break;
+                }
+            }
+        }
+        block
+    }
+
+    #[ignore]
+    #[test]
+    fn add_vote() {
+        let nodes = create_full_infos();
+        let valid_voters = nodes.iter().map(FullInfo::public_info).collect::<Vec<_>>();
+        let mut accumulator = VoteAccumulator::default();
+
+        // Add votes for the same payload repeatedly.  Until we get to quorum, an empty vec should
+        // should be returned.  On reaching quorum, we should have a single entry returned
+        // containing the now-valid block.  Subsequent votes should be returned as single entries.
+        // Also have each node double-vote (i.e. vote for the same thing twice) for a second payload
+        // to ensure that the behaviour matches as for the first, i.e. that identical votes have no
+        // effect.
+        let payload1 = "Live";
+        let payload2 = "DoubleVote";
+        let mut is_valid = false;
+        let mut nodes_added = vec![];
+        for node in &nodes {
+            let vote1 = unwrap!(Vote::new(node.secret_sign_key(), payload1));
+            let vote2 = unwrap!(Vote::new(node.secret_sign_key(), payload2));
+            let results1 = unwrap!(accumulator.add_vote(
+                vote1,
+                node.public_info(),
+                valid_voters.clone(),
+            ));
+            let results2 = unwrap!(accumulator.add_vote(
+                vote2,
+                node.public_info(),
+                valid_voters.clone(),
+            ));
+            nodes_added.push(node.public_info());
+
+            if data_chain::quorum(valid_voters.clone(), nodes_added.iter().cloned()) {
```

Comment:
Shouldn't the arguments be reversed? That is, `nodes_added` first, `valid_voters` second?